File: scalability/fode_scalability.py
#!/usr/bin/env python3
"""Run one pycaputo PECE dimension/interval scalability benchmark.

Usage:
    python3 scalability/fode_scalability.py D T N
    python3 scalability/fode_scalability.py --baseline

The benchmark is D_t^0.8 y = A y with the same matrix, initial state, time
grid, and one PECE corrector iteration used by the Julia and MATLAB workers.
"""

from __future__ import annotations

import gc
import os
import sys
import time

os.environ.setdefault("OMP_NUM_THREADS", "1")
os.environ.setdefault("OPENBLAS_NUM_THREADS", "1")
os.environ.setdefault("MKL_NUM_THREADS", "1")

import numpy as np
from pycaputo.controller import FixedController
from pycaputo.derivatives import CaputoDerivative
from pycaputo.events import StepCompleted
from pycaputo.fode import caputo
from pycaputo.stepping import evolve


ALPHA = 0.8
T0 = 0.0


def system_matrix(dimension: int) -> np.ndarray:
    return (
        -1.5 * np.eye(dimension, dtype=float)
        + 0.25 * np.eye(dimension, k=1, dtype=float)
        + 0.25 * np.eye(dimension, k=-1, dtype=float)
    )


def initial_state(dimension: int) -> np.ndarray:
    return np.linspace(1.0, 0.1, dimension, dtype=float)


def solve_pece(
    dimension: int, final_time: float, number_of_steps: int
) -> tuple[np.ndarray, np.ndarray, float]:
    h = (final_time - T0) / number_of_steps
    matrix = system_matrix(dimension)

    def source(_t: float, y: np.ndarray) -> np.ndarray:
        return matrix @ y

    stepper = caputo.PECE(
        ds=tuple(CaputoDerivative(ALPHA) for _ in range(dimension)),
        # Stop by N rather than tfinal. Supplying tfinal makes pycaputo add a
        # small epsilon to later steps, so the resulting mesh is not exactly h.
        control=FixedController(
            tstart=T0,
            tfinal=None,
            nsteps=number_of_steps,
            dt=h,
        ),
        source=source,
        y0=(initial_state(dimension),),
        corrector_iterations=1,
    )

    times: list[float] = []
    values: list[np.ndarray] = []
    for event in evolve(stepper, dtinit=h):
        if isinstance(event, StepCompleted):
            times.append(float(event.t))
            values.append(np.array(event.y, copy=True))

    if not values:
        raise RuntimeError("pycaputo PECE returned no completed steps")

    times_array = np.asarray(times)
    values_array = np.stack(values)
    if times_array.size != number_of_steps + 1:
        raise RuntimeError("pycaputo PECE returned an unexpected number of points")
    if values_array.shape != (number_of_steps + 1, dimension):
        raise RuntimeError("pycaputo PECE returned an unexpected state shape")
    tolerance = (
        8
        * number_of_steps
        * np.finfo(float).eps
        * max(1.0, abs(T0), abs(final_time))
    )
    if not np.isclose(times_array[-1], final_time, rtol=0.0, atol=tolerance):
        raise RuntimeError("pycaputo PECE did not reach the requested final time")

    return times_array, values_array, h


def parse_positive_int(value: str, name: str) -> int:
    try:
        parsed = int(value)
    except ValueError as exc:
        raise SystemExit(f"{name} must be a positive integer") from exc
    if parsed <= 0:
        raise SystemExit(f"{name} must be positive")
    return parsed


def parse_positive_float(value: str, name: str) -> float:
    try:
        parsed = float(value)
    except ValueError as exc:
        raise SystemExit(f"{name} must be a positive number") from exc
    if not np.isfinite(parsed) or parsed <= T0:
        raise SystemExit(f"{name} must be finite and positive")
    return parsed


def main(argv: list[str]) -> None:
    if argv[1:] == ["--baseline"]:
        print("SCALABILITY_RESULT,0,0,0,0,0", flush=True)
        return
    if len(argv) != 4:
        raise SystemExit("usage: fode_scalability.py D T N | --baseline")

    dimension = parse_positive_int(argv[1], "D")
    final_time = parse_positive_float(argv[2], "T")
    number_of_steps = parse_positive_int(argv[3], "N")
    h = (final_time - T0) / number_of_steps

    warmup_steps = min(number_of_steps, 64)
    warmup_final_time = T0 + warmup_steps * h
    warmup_times, warmup_values, _ = solve_pece(
        dimension, warmup_final_time, warmup_steps
    )
    del warmup_times, warmup_values
    gc.collect()

    start = time.perf_counter()
    times, values, returned_h = solve_pece(dimension, final_time, number_of_steps)
    runtime_seconds = time.perf_counter() - start
    if returned_h != h:
        raise RuntimeError("internal step-size mismatch")
    if not np.all(np.isfinite(values[-1])):
        raise RuntimeError("pycaputo PECE returned a non-finite solution")

    # Retain the returned trajectory until after the RSS sample/result output.
    _trajectory_size = times.nbytes + values.nbytes
    print(
        "SCALABILITY_RESULT,"
        f"{dimension},{final_time:.17g},{number_of_steps},"
        f"{h:.17g},{runtime_seconds:.17g}",
        flush=True,
    )
    del _trajectory_size


if __name__ == "__main__":
    main(sys.argv)

File: scalability/fode_scalability.jl
#!/usr/bin/env julia

"""
Run one Julia PECE solve for the dimension/interval scalability benchmark.

Usage:
    julia --project=. scalability/fode_scalability.jl D T N
    julia --project=. scalability/fode_scalability.jl --baseline

The benchmark is the commensurate system

    D_t^0.8 u(t) = A_d u(t),  u(0) = [1.0, ..., 0.1],

where A_d has -1.5 on the diagonal and 0.25 on its first off-diagonals.
The final stdout line is parsed by `scalability_figure.jl`.
"""

using FractionalDiffEq
using LinearAlgebra
using Printf

const ALPHA = 0.8
const T0 = 0.0

function system_matrix(dimension::Int)
    matrix = -1.5 * Matrix{Float64}(I, dimension, dimension)
    for index in 1:(dimension - 1)
        matrix[index, index + 1] = 0.25
        matrix[index + 1, index] = 0.25
    end
    return matrix
end

initial_state(dimension::Int) = collect(range(1.0, 0.1; length = dimension))

function rhs!(du, u, matrix, _t)
    mul!(du, matrix, u)
    return nothing
end

function solve_pece(dimension::Int, final_time::Float64, number_of_steps::Int)
    h = (final_time - T0) / number_of_steps
    orders = fill(ALPHA, dimension)
    problem = FODEProblem(
        ODEFunction{true}(rhs!),
        orders,
        initial_state(dimension),
        (T0, final_time),
        system_matrix(dimension),
    )
    return solve(problem, PECE(); dt = h), h
end

function parse_positive_int(value::String, name::String)
    parsed = try
        parse(Int, value)
    catch
        error("$(name) must be a positive integer; received $(repr(value))")
    end
    parsed > 0 || error("$(name) must be positive")
    return parsed
end

function parse_positive_float(value::String, name::String)
    parsed = try
        parse(Float64, value)
    catch
        error("$(name) must be positive; received $(repr(value))")
    end
    isfinite(parsed) && parsed > T0 || error("$(name) must be finite and positive")
    return parsed
end

function main(args)
    if args == ["--baseline"]
        @printf("SCALABILITY_RESULT,0,0,0,0,0\n")
        return
    end
    length(args) == 3 || error("usage: fode_scalability.jl D T N | --baseline")
    dimension = parse_positive_int(args[1], "D")
    final_time = parse_positive_float(args[2], "T")
    number_of_steps = parse_positive_int(args[3], "N")
    h = (final_time - T0) / number_of_steps

    # Compile the same solver path using the requested h over at most 64 steps.
    warmup_steps = min(number_of_steps, 64)
    warmup_final_time = T0 + warmup_steps * h
    warmup_solution, _ = solve_pece(dimension, warmup_final_time, warmup_steps)
    isempty(warmup_solution.u) && error("Julia PECE warm-up returned no values")
    warmup_solution = nothing
    GC.gc(true)

    solution = nothing
    runtime_seconds = @elapsed begin
        solution, returned_h = solve_pece(dimension, final_time, number_of_steps)
        returned_h == h || error("internal step-size mismatch")
    end

    length(solution.t) == number_of_steps + 1 ||
        error("Julia PECE returned an unexpected number of time points")
    isapprox(last(solution.t), final_time; rtol = 0, atol = 64eps(final_time)) ||
        error("Julia PECE did not reach the requested final time")
    final_state = last(solution.u)
    length(final_state) == dimension || error("Julia PECE returned an invalid state size")
    all(isfinite, final_state) || error("Julia PECE returned a non-finite solution")

    @printf(
        "SCALABILITY_RESULT,%d,%.17g,%d,%.17g,%.17g\n",
        dimension,
        final_time,
        number_of_steps,
        h,
        runtime_seconds,
    )
end

main(ARGS)
